### PR TITLE
sys/hashes/pbkdf2: Add PBKDF2-sha256 implementation. [TAKEOVER]

### PR DESCRIPTION
**Pull request by fjmolinas**
### Contribution description

This PR rebases and fixes murdock complaints for  #12211, I also fixed some issues when running the test on hardware. I took over since I can't push to the archived repository.

This add an implementation of PBKDF2 using sha256 hmac. Only one derived key length is supported (32) though it should not be hard to extend it.

### Testing procedure

The testing is done with both random (with fixed seed) vectors and vectors from rfc7914. 

<details><summary>BOARD=native make -C tests/pbkdf2/ all test -j3 --no-print-directory</summary>

```
Building application "tests_pbkdf2" for "native" with MCU "native".

"make" -C /home/francisco/workspace/RIOT3/boards/native
"make" -C /home/francisco/workspace/RIOT3/boards/native/drivers
"make" -C /home/francisco/workspace/RIOT3/core
"make" -C /home/francisco/workspace/RIOT3/cpu/native
"make" -C /home/francisco/workspace/RIOT3/cpu/native/periph
"make" -C /home/francisco/workspace/RIOT3/drivers
"make" -C /home/francisco/workspace/RIOT3/drivers/periph_common
"make" -C /home/francisco/workspace/RIOT3/cpu/native/stdio_native
"make" -C /home/francisco/workspace/RIOT3/sys
"make" -C /home/francisco/workspace/RIOT3/sys/auto_init
"make" -C /home/francisco/workspace/RIOT3/sys/base64
"make" -C /home/francisco/workspace/RIOT3/sys/crypto
"make" -C /home/francisco/workspace/RIOT3/sys/hashes
"make" -C /home/francisco/workspace/RIOT3/sys/test_utils/interactive_sync
   text    data     bss     dec     hex filename
  35234     708   48228   84170   148ca /home/francisco/workspace/RIOT3/tests/pbkdf2/bin/native/tests_pbkdf2.elf
r
make[1]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
/home/francisco/workspace/RIOT3/tests/pbkdf2/bin/native/tests_pbkdf2.elf
RIOT native interrupts/signals initialized.
LED_RED_OFF
LED_GREEN_ON
RIOT native board initialized.
RIOT native hardware initialization complete.

Help: Press s to start test, r to print it is ready
READY
s
START
main(): This is RIOT! (Version: 2021.01-devel-84-gb2295-pbkdf2-sha256)
{ready}
passwd
{ready}
c2FsdA==
{ready}
1
VawEblbjCJ/sFpHCJUS2BflBhSFt3gRl5oudV8INrLw=
{ready}
Password
{ready}
TmFDbA==
{ready}
80000
TdzY9guYviGDDO5e8icB+WQaRBjQTAQUrv8Ih2s0q1Y=
{ready}

r
make[1]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
/home/francisco/workspace/RIOT3/tests/pbkdf2/bin/native/tests_pbkdf2.elf
RIOT native interrupts/signals initialized.
LED_RED_OFF
LED_GREEN_ON
RIOT native board initialized.
RIOT native hardware initialization complete.

Help: Press s to start test, r to print it is ready
READY
s
START
main(): This is RIOT! (Version: 2021.01-devel-84-g4875a-pbkdf2-sha256)
{ready}

{ready}
vdZA+wZnGtEcgDF/o7F5nQ==
{ready}
10
mkkhkWNNxh8SRegffJ3jT+/INJDwceUQ/JGGAXZJ1mg=
{ready}
:5e`t(H"
{ready}

{ready}
10
HJ8E5DsSemPdwcEcdiWsXNHUNK7FpfL6NPqE2Zl0Qho=
{ready}
4P"2]S4Wl
{ready}
YULqfRe+MREaKnPtVisPecN0We71C+pjNx7NeyfNgTBHIpOJVxqodmwwdRGyuUN6KN9uxM5KK73CQTMLAannHw==
{ready}
1
7/kTA8PIj0TXUHE4i5VCbre7T5b0sDI8JizsR8cH3Wg=
{ready}
)pYleR|CTnZqVb$5;'6):\BC39x]Y0e/D~\Tapi5#=94ys=^EvK97U8WuE4~P($*[
{ready}
d9L/Ac+ZmIwkyWGxzSJigBxFEENaEJiuQzRsEqzorjQEVMrFtow=
{ready}
20
LVbkmVr4rwFkBWxVg76nwLSh6dYmdOCuw2KTgYRkZfE=
{ready}
R|q!d`S9\*IKzs9O0vr<\Y.hSiR >!xs
{ready}
PUy/Trk+/86Iyy3U6Ag5
{ready}
12
Y2T4na5rpTpEsIViaVnINwG0R9CwhPl0ta6myAaBtIs=
{ready}
%sy(N&hh,MT9rH4Se3=~]IQ+5@W54&[5vq&6_4,xVLjl2)HH
{ready}
pWwJQfvyQFCnSNvPrGGeYw3eKaa6pLca3SRnrHeO7bM=
{ready}
15
txemIhZNx1TYOd+Z1UIpCrtPy469VTitUNNO+QX8YFU=
{ready}

```
</details>

<details><summary>BOARD=nrf52840-mdk make -C tests/pbkdf2/ flash test -j3 --no-print-directory</summary>

```
main(): This is RIOT! (Version: 2021.01-devel-84-g4875a-pbkdf2-sha256)
{ready}
passwd
{ready}
c2FsdA==
{ready}
1
VawEblbjCJ/sFpHCJUS2BflBhSFt3gRl5oudV8INrLw=
{read
r
make[1]: warning: jobserver unavailable: using -j1.  Add '+' to parent make rule.
socat - open:/dev/riot/tty-nrf52840-mdk,b115200,echo=0,raw
Help: Press s to start test, r to print it is ready
READY
s
START
main(): This is RIOT! (Version: 2021.01-devel-84-g4875a-pbkdf2-sha256)
{ready}

{ready}
vdZA+wZnGtEcgDF/o7F5nQ==
{ready}
10
mkkhkWNNxh8SRegffJ3jT+/INJDwceUQ/JGGAXZJ1mg=
{ready}
:5e`t(H"
{ready}

{ready}
10
HJ8E5DsSemPdwcEcdiWsXNHUNK7FpfL6NPqE2Zl0Qho=
{ready}
4P"2]S4Wl
{ready}
YULqfRe+MREaKnPtVisPecN0We71C+pjNx7NeyfNgTBHIpOJVxqodmwwdRGyuUN6KN9uxM5KK73CQTMLAannHw==
{ready}
1
7/kTA8PIj0TXUHE4i5VCbre7T5b0sDI8JizsR8cH3Wg=
{ready}
)pYleR|CTnZqVb$5;'6):\BC39x]Y0e/D~\Tapi5#=94ys=^EvK97U8WuE4~P($*[
{ready}
d9L/Ac+ZmIwkyWGxzSJigBxFEENaEJiuQzRsEqzorjQEVMrFtow=
{ready}
20
LVbkmVr4rwFkBWxVg76nwLSh6dYmdOCuw2KTgYRkZfE=
{ready}
R|q!d`S9\*IKzs9O0vr<\Y.hSiR >!xs
{ready}
PUy/Trk+/86Iyy3U6Ag5
{ready}
12
Y2T4na5rpTpEsIViaVnINwG0R9CwhPl0ta6myAaBtIs=
{ready}
%sy(N&hh,MT9rH4Se3=~]IQ+5@W54&[5vq&6_4,xVLjl2)HH
{ready}
pWwJQfvyQFCnSNvPrGGeYw3eKaa6pLca3SRnrHeO7bM=
{ready}
15
txemIhZNx1TYOd+Z1UIpCrtPy469VTitUNNO+QX8YFU=
{ready}
```
</details>

### Issues/PRs references

Taken from #12191 .
Closes #12211

### CODE REVIEW

File: sys/hashes/pbkdf2.c [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    int len = SHA256_DIGEST_LENGTH;
+
+    while (len--) {
+        *d1 ^= *d2;
+        d1++;
+        d2++;
+    }
+}
+
+#define _WIPE(a) crypto_secure_wipe(&a, sizeof(a))
```
Comment (benpicco) [inline sys/hashes/pbkdf2.c:46]: That define obscures more than it adds.

File: sys/include/hashes/pbkdf2.h [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+#define PBKDF2_KEY_SIZE SHA256_DIGEST_LENGTH
+
+/**
+ * @brief Create a key from a password and hash using PBKDF2.
+ *
+ * @param[in]   password        password pointer
+ * @param[in]   password_len    length of password
+ * @param[in]   salt            salt pointer
+ * @param[in]   salt_len        salt length
+ * @param[in]   iterations      number of rounds. Must be >1.
```
Comment (benpicco) [inline sys/include/hashes/pbkdf2.h:45]: Can we recommend some value here?

**Review by benpicco — COMMENTED** [review]
looks good, you want to add those four `atmega328p` to `Makefile.ci`.

Also the automatic test failed on CI. Does it work locally?

**Comment by fjmolinas** [comment]
> Also the automatic test failed on CI. Does it work locally?

It did... I'll take a look..

File: sys/hashes/pbkdf2.c (hunk 1 above)
Comment (fjmolinas) [reply sys/hashes/pbkdf2.c:46]: Should I replace by the function call?

Comment (benpicco) [reply sys/hashes/pbkdf2.c:46]: Yes

File: sys/include/hashes/pbkdf2.h (hunk 2 above)
Comment (fjmolinas) [reply sys/include/hashes/pbkdf2.h:45]: Indeed although the recommended value is not staright forward as nicely described in https://security.stackexchange.com/questions/3959/recommended-of-iterations-when-using-pkbdf2-sha256 and https://cryptosense.com/blog/parameter-choice-for-pbkdf2/, maybe 10000 would be a good recommendation based on these.

**Comment by fjmolinas** [comment]
> It did... I'll take a look..

Ahh its a python version issue.

File: tests/pbkdf2/tests/02-random.py [hunk 3]
```diff
@@ -18,6 +18,29 @@
...
+        cum_weights = list(_accumulate(weights))
+    elif weights is not None:
+        raise TypeError('Cannot specify both weights and cumulative weights')
+    if len(cum_weights) != n:
+        raise ValueError('The number of weights does not match the population')
+    bisect = _bisect
+    total = cum_weights[-1] + 0.0   # convert to float
+    hi = n - 1
+    return [population[bisect(cum_weights, random.random() * total, 0, hi)]
+            for i in _repeat(None, k)]
```
Comment (fjmolinas) [inline tests/pbkdf2/tests/02-random.py:43]: well this was added in the wrong place hahahaha...

Comment (fjmolinas) [reply tests/pbkdf2/tests/02-random.py]: Lets see if now it works

**Comment by fjmolinas** [comment]
This time the test passed, there where only some unrelated failures

```
--- run_test job results (2 failed, 230 passed, 232 total):
    failed:
    tests/mtd_flashpage/samr21-xpro:llvm
    tests/pkg_libhydrogen/samr21-xpro:llvm
```

There was some boards that needed to be added to the insufficient memory list, as well as a typo to be fixed. Is it OK to squash @benpicco ?

**Comment by benpicco** [comment]
Sure, please squash - but also go recommend some parameter lengths in the doc so it's clear how to use this properly

File: sys/include/hashes/pbkdf2.h (hunk 2 above)
Comment (fjmolinas) [reply sys/include/hashes/pbkdf2.h:45]: @benpicco I already added the recommendation here!

Comment (benpicco) [reply sys/include/hashes/pbkdf2.h:45]: Ah sorry, didn't see that

**Comment by benpicco** [comment]
Only `run_test/tests/pkg_libhydrogen/samr21-xpro:llvm` failed as expected.

**Review by benpicco — APPROVED** [review]

**Pull request merged**